### PR TITLE
fix(Core): Crashfix.

### LINKED ISSUE #10965: Crash bind to reset instance
### Current Behaviour

A crash server come ofen after a reset instance. 

### Expected Blizzlike Behaviour

None

### Source

The source seems to have been located by @SilentCLD (sources #10819 ) : 

Screenshot 2022-03-10 145234

Dump file:
[6e35325a9cef_worldserver.exe_[10-3_8-30-50].zip](https://github.com/azerothcore/azerothcore-wotlk/files/8221231/6e35325a9cef_worldserver.exe_.10-3_8-30-50.zip)

Worldserver error:
Screenshot 2022-03-10 083129

### Steps to reproduce the problem

1. binding to an instance
2. gracefully restarting the server
3. instance unbind all

### Extra Notes

_No response_

### AC rev. hash/commit

AzerothCore rev. 6e35325a9cef

### Operating system

Windows 11

### Custom changes or Modules

_No response_

### PR DESCRIPTION
**Pull request by UltraNix**
Fixes #10965

## Issues Addressed:

- Closes #10965

## Tests Performed:

- Not tested.

## How to Test the Changes:

More info [here](https://github.com/azerothcore/azerothcore-wotlk/issues/10965)

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

File: src/server/game/Maps/Map.cpp [hunk 1]
```diff
@@ -2797,6 +2797,7 @@ InstanceMap::~InstanceMap()
 {
     delete instance_data;
     instance_data = nullptr;
+    sInstanceSaveMgr->DeleteInstanceSaveIfNeeded(GetInstanceId(), true);
```
Comment (UltraNix) [inline src/server/game/Maps/Map.cpp:2800]: @Kitzunu https://github.com/azerothcore/azerothcore-wotlk/commit/f9d708b45005cf6327e3a73d2062a2f92bbe9de6#diff-cacd2443c0f46ce614d849924bdb64dd81aa4d5068b131e1c71ecac518b1c3e1L2615
Why this line has been removed?

**Comment by UltraNix** [comment]
@Lynethris @Nefertumm @DepTypes Please, test it.

**Comment by DepTypes** [comment]
`Unhandled exception thrown: read access violation.`

```
 	worldserver.exe!std::list<ObjectGuid,std::allocator<ObjectGuid>>::begin() Line 1075	C++
>	worldserver.exe!InstanceSaveMgr::UnbindAllFor(InstanceSave * save) Line 807	C++
 	worldserver.exe!Group::ResetInstances(unsigned char method, bool isRaid, Player * leader) Line 2049	C++
 	worldserver.exe!WorldSession::HandleResetInstancesOpcode(WorldPacket & __formal) Line 1267	C++
```

**Comment by Lynethris** [comment]
Tested and no error for me (OS : Windows 11, Tools: VS2022): 

- [X] No error during compiling
- [X] Test with the method #10965 
- [X] Different test with restart server (gm commands), force shutdown (just closing the windows), different reset instance at different moments

**Comment by DepTypes** [comment]
- I have 2 characters at the entrance to BRD,.
- Kill one NPC
- Run out
- `!server restart 1`
- Log back in with both characters, the server crashes when you reset the instance

2 chracters are required for the crash because the InstanceSave for all characters is deleted before all characters have been removed from it causing the ptr to point to garbage.

File: src/server/game/Instances/InstanceSaveMgr.cpp [hunk 2]
```diff
@@ -136,7 +136,11 @@ bool InstanceSaveMgr::DeleteInstanceSaveIfNeeded(InstanceSave* save, bool skipMa
         // clear respawn times (if map is loaded do it just to be sure, if already unloaded it won't do it by itself)
         Map::DeleteRespawnTimesInDB(save->GetMapId(), save->GetInstanceId());
 
-        delete save;
+        if (deleteSave)
+        {
+            delete save;
+        }
+
         return true;
```
Comment (DepTypes) [inline src/server/game/Instances/InstanceSaveMgr.cpp:144]: should `return true` right after the delete. The return is on the wrong side of the brace and always returns true.

**Comment by UltraNix** [comment]
Still need some testing as @DepTypes said it's still crashing.

**Comment by Lynethris** [comment]
> Still need some testing as @DepTypes said it's still crashing.

Confirmed. I made the mistake to have 2 char in the same group.

**Comment by UltraNix** [comment]
@DepTypes @Lynethris Update. Please test it again.

**Comment by DepTypes** [comment]
_Don't be so hasty..._ I need to remind myself. Nevermind the above, the problem was

```cpp
void InstanceSaveMgr::UnbindAllFor(InstanceSave* save)
{
    GuidList& pList = save->m_playerList;
    while (!pList.empty())
        PlayerUnbindInstance(*(pList.begin()), save->GetMapId(), save->GetDifficulty(), true, ObjectAccessor::FindConnectedPlayer(*(pList.begin())));
}
```
calling map->m_playerList after the map is deleted.
--

Pulling and testing again.

**Comment by DepTypes** [comment]
Ran the same test procedure twice.
No crash and database is now cleared of saves as expected. ✔️

**Comment by Lynethris** [comment]
Tested and work. 

Global method: 
- 2 characters at the entrance to BRD,.
- Kill one NPC
- Run out
- !server restart 1

3 situations tested (repeat 2 times):
- 2 char in the same, and enter together 
- 2 char in the same group, but only one enter in the dungeon
- 2 char online, but not in group. Only one enter in the dungeon

**Review by DepTypes — APPROVED** [review]
Solves the crashes. The critical issue is solved with this PR.

**Review by Nefertumm — APPROVED** [review]

**Pull request merged**